Use scaling factor to compute ttnn.mean for multi-dim non w,h case + scaling factor fixes  (#28644)

yiding opened the pull request:
### Ticket

#28644

### Problem description
Remove ttnn.mul_sfpu call when computing multidimensional ttnn.mean on dimensions that are not the last 2 dimensions.

### What's changed

1. Fixed `reduce_nd_loop` applying scalars per axis reduction, and instead only apply it on the first axis. This corrects the issue that multipliers are effectively applied multiple times, once per reduction axis. Note: that behavior of `scalar` param is still sometimes incorrect when hitting other code paths (for example, ttnn.sum with a non -1, -2 dimension will call code that ignores scalar param).

2. Fixed scaling factor computation for multi-dimensional non w,h mean reduction. This used to use the volume of the input tensor, which is only correct for reducing to scalar tensors. Added a test to check this, existing tests do not cover this as assert_with_pcc ignores differences by constant factors.

3. The original ticket: multi-dim non w,h mean scaling is performed in `reduce_nd_loop`  instead of a separate mul_sfpu call.

### Perf

The performance improvement should be most apparent on reductions on small axis. Test was performed on P150. The effect of mul_sfpu seems small in even this worst-case scenario, however the traces demonstrate that it has indeed been eliminated.

Aggregate report from tracy:

*shape=(2, 2, 1024, 1024) dim=(0, 1)*

Before:
```
    %                                OP Code Joined  Device_Time_Sum_us  Ops_Count
52.21              Transpose (in0:dram_interleaved)            13251.46         10
25.57 PermuteDeviceOperation (in0:dram_interleaved)             6490.56         10
21.83                 Reduce (in0:dram_interleaved)             5539.40         10
 0.39   UnaryDeviceOperation (in0:dram_interleaved)               99.17          5
```

After:
```
    %                                OP Code Joined  Device_Time_Sum_us
52.29              Transpose (in0:dram_interleaved)            13232.29
25.81 PermuteDeviceOperation (in0:dram_interleaved)             6531.49
21.90                 Reduce (in0:dram_interleaved)             5542.71
```


*shape=(2, 2, 4096, 4096) dim=(0, 1)*

Before:
```
    %                                OP Code Joined  Device_Time_Sum_us  Ops_Count
52.56              Transpose (in0:dram_interleaved)           210945.69         10
25.54 PermuteDeviceOperation (in0:dram_interleaved)           102496.66         10
21.67                 Reduce (in0:dram_interleaved)            86981.73         10
 0.23   UnaryDeviceOperation (in0:dram_interleaved)              925.89          5
```

After:
```
    %                                OP Code Joined  Device_Time_Sum_us  Ops_Count
52.53              Transpose (in0:dram_interleaved)           209601.36         10
25.63 PermuteDeviceOperation (in0:dram_interleaved)           102254.18         10
21.84                 Reduce (in0:dram_interleaved)            87148.82         10
```


Test code:
```
@pytest.mark.parametrize("shape", [(2, 2, 1024, 1024), (2, 2, 4096, 4096)])
@pytest.mark.parametrize("dim", [[0, 1]])
def test_benchmark_perf_mean_nd(device, shape, dim):
    import tracy

    # best case for mul_sfpu removal.
    torch.manual_seed(0)
    torch_input_tensor = torch_random(shape, -1, 1, dtype=torch.bfloat16)

    tracy.signpost(f"warmup-{"-".join(str(x) for x in shape)}")
    # warmup
    input_tensor = ttnn.from_torch(torch_input_tensor, layout=ttnn.TILE_LAYOUT, device=device)
    output_tensor = ttnn.mean(input_tensor, dim=dim, keepdim=True)

    tracy.signpost(f"test-{"-".join(str(x) for x in shape)}")
    # test
    for _ in range(5):
        output_tensor = ttnn.mean(input_tensor, dim=dim, keepdim=True)
        output_tensor = ttnn.to_torch(output_tensor)
```


### Checklist
- [x] test_reduction.py tests pass
- [x] test_reduction_mean.py tests pass

- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

nsorabaTT reviewed (changes requested):
Thanks for working on this bounty. Overall, the PR is in a good shape. I have added two minor comments below. Please address them and I will approve.

We will also need the pipelines green before the merge. I will kick them off.

nsorabaTT commented:
> Thanks for working on this bounty. Overall, the PR is in a good shape. I have added two minor comments below. Please address them and I will approve.
> 
> We will also need the pipelines green before the merge. I will kick them off.

Please rebase your branch on top of the latest changes in main and resolve conflicts (if any).

yiding commented on tests/ttnn/unit_tests/operations/reduce/test_reduction_mean.py:
```diff
@@ -28,3 +28,22 @@ def test_mean(device, batch_size, h, w, dim, keepdim):
+    input_tensor = ttnn.from_torch(torch_input_tensor, layout=ttnn.TILE_LAYOUT, device=device)
+    ttnn.fill_implicit_tile_padding(input_tensor, 42)  # garbage padding to test that mean removes it
+
+    output_tensor = ttnn.mean(input_tensor, dim=dim, keepdim=keepdim)
```
I can, but there's another code path where scalar is still broken so I didn't do it. I can add the tests and disable the broken cases and file a bug for that.

yiding committed:
prevent reduce op scalar multiplier from being applied once per axis for axis-by-axis reductions

yiding committed:
fix ttnn.mean axis-by-axis implementation to compute correct scaling factor

yiding committed:
ttnn.mean: use reduce scalar parameter instead of mul_sfpu for reduce_nd_loop case

yiding commented:
Rebased, pulled out the scalar computation, and added a (disabled) test for scalar.

yiding commented:
That's one of the tests I added that is averaging ones(), so seed shouldn't matter. FWIW I am using blackhole and test seems like it's on wormhole.

yiding commented:
I fixed my bug by putting back the bool first = ... check, since I think it's simpler than trying to intuit whether the dimension is the first dimension in the list of dimensions being reduced.

Also I can re-enable the new test I added for non-unity scaling factor, as the problem I was thinking of only apply to sum and not mean.
